File: 백준/그래프 이론/Watering the Fields.py
"""
시작 시간: 2022년 7월 1일 오전 2시
소요 시간: 35분
풀이 방법:
    시간초과!
    -> c++로 똑같이 작성하니 통과됨. 파이썬 느린거 채점시스템에서 추가시간 안줘서 문제난듯
"""
import sys
input = sys.stdin.readline

def get_cost(point1, point2):
    return (point1[0]-point2[0])**2 + (point1[1]-point2[1])**2

def get_edges():
    global parents, heights
    n, c = map(int, input().split())
    parents = [i for i in range(n)]
    heights = [0] * n
    edges = []
    vertexes = []
    for i in range(n):
        v = tuple(map(int, input().split()))
        for j, other in enumerate(vertexes):
            cost = get_cost(v, other)
            edges.append((cost, i, j))
        vertexes.append(v)
    edges.sort()
    return edges, c # 엣지랑 최소 값

def find(v):
    global parents
    while v != parents[v]:
        v = parents[v]
    return v

def union(u, v):
    global parents, heights
    u, v = find(u), find(v)
    if u == v: return
    if heights[u] > heights[v]:
        parents[v] = u
    else:
        parents[u] = v
        if heights[u] == heights[v]:
            heights[v] += 1

# 크루스칼 시작
edges, min_cost = get_edges()
ret = 0
i = 0
for i, (cost, v, u) in enumerate(edges):
    if find(v) == find(u): continue
    if cost < min_cost: continue
    union(v, u)
    ret += cost

group = find(0)
is_successed = True
for i in range(len(parents)):
    if group != find(i):
        is_successed = False
        break
if is_successed:
    print(ret)
else:
    print(-1)

File: 백준/그래프 이론/Watering the Fields.cpp
/*
    크루스칼 사용: 파이썬 코드와 똑같은 논리전개
*/
#include <iostream>
#include <queue>
using namespace std;
struct edge
{
    int here, other, cost;
};
struct compare
{
    bool operator()(edge &i, edge &o)
    {
        return i.cost > o.cost;
    }
};

int get_cost(pair<int, int> point1, pair<int, int> point2)
{
    int dx = point1.first - point2.first;
    int dy = point1.second - point2.second;
    return dx * dx + dy * dy;
}

priority_queue<edge, vector<edge>, compare> get_edges(int n)
{
    priority_queue<edge, vector<edge>, compare> edges;
    pair<int, int> *vertexes = new pair<int, int>[n];
    for (int i = 0; i < n; i++)
    {
        pair<int, int> p;
        cin >> p.first >> p.second;
        vertexes[i] = p;
        for (int j = 0; j < i; j++)
        {
            edge e;
            e.cost = get_cost(vertexes[i], vertexes[j]);
            e.here = i;
            e.other = j;
            edges.push(e);
        }
    }
    return edges;
}

int *parents, *heights;
int find(int v)
{

    while (v != parents[v])
        v = parents[v];
    return v;
}
void merge(int v, int u) // union 함수
{
    u = find(u);
    v = find(v);
    if (u == v)
        return;
    if (heights[u] > heights[v])
        parents[v] = u;
    else
    {
        parents[u] = v;
        if (heights[u] == heights[v])
            heights[v]++;
    }
}

int main()
{
    priority_queue<edge, vector<edge>, compare> edges;
    int n, c; // 정점 개수, 최소 비용
    cin >> n >> c;
    parents = new int[n];
    for (int i = 0; i < n; i++)
        parents[i] = i;
    heights = new int[n]();
    edges = get_edges(n);

    /* 크루스칼 시작 */
    unsigned int ret = 0;
    while (!edges.empty())
    {
        edge e = edges.top();
        edges.pop();
        if (find(e.here) == find(e.other))
            continue;
        if (e.cost < c)
            continue;
        merge(e.here, e.other);
        ret += e.cost;
    }

    int group = find(0);
    bool is_successed = true;
    for (int i = 0; i < n; i++)
    {
        if (group != find(i))
        {
            is_successed = false;
            break;
        }
    }
    if (is_successed)
        cout << ret;

    else
        cout << -1;
    return 0;
}